# HG changeset patch
# User andrew
# Date 1345037736 -3600
# Node ID 35e024c6a62cdd413c32ce64e9dfd82900dda0cd
# Parent  f0bf7358ba23e5554e0e1cef3e9705aefb1a0e91
7110151: Use underlying platform's zlib library for Java zlib support
Summary: Make SYSTEM_ZLIB more flexible by using ZLIB_{CFLAGS,LIBS} and building on more than just MACOSX.
Reviewed-by: sherman, alanb

diff --git a/make/com/sun/java/pack/Makefile b/make/com/sun/java/pack/Makefile
--- jdk8/jdk/make/com/sun/java/pack/Makefile
+++ jdk8/jdk/make/com/sun/java/pack/Makefile
@@ -77,8 +77,8 @@
   OTHER_CXXFLAGS += $(ZINCLUDE)
   LDDFLAGS += $(ZIPOBJS)
  else
-  LDDFLAGS += -lz
-  OTHER_CXXFLAGS += -DSYSTEM_ZLIB
+  LDDFLAGS += $(ZLIB_LIBS)
+  OTHER_CXXFLAGS += $(ZLIB_CFLAGS) -DSYSTEM_ZLIB
  endif
 else
   OTHER_CXXFLAGS += -DNO_ZLIB -DUNPACK_JNI
diff --git a/make/common/Program.gmk b/make/common/Program.gmk
--- jdk8/jdk/make/common/Program.gmk
+++ jdk8/jdk/make/common/Program.gmk
@@ -91,7 +91,7 @@
   LDFLAGS += $(OUTPUTDIR)/tmp/java/jli/$(OBJDIRNAME)/static/libjli.a
 
   ifeq ($(SYSTEM_ZLIB),true)
-    OTHER_LDLIBS += -lz
+    OTHER_LDLIBS += $(ZLIB_LIBS)
   endif
 endif
 
diff --git a/make/common/shared/Defs-linux.gmk b/make/common/shared/Defs-linux.gmk
--- jdk8/jdk/make/common/shared/Defs-linux.gmk
+++ jdk8/jdk/make/common/shared/Defs-linux.gmk
@@ -129,6 +129,11 @@
 BUILD_HEADLESS = true
 LIBM=-lm
 
+# Set ZLIB_LIBS if not already set
+ifeq ("$(ZLIB_LIBS)", "")
+  ZLIB_LIBS=-lz
+endif
+
 # GCC29_COMPILER_PATH: is the path to where the gcc 2.9 compiler is installed
 #  NOTE: Must end with / so that it could be empty, allowing PATH usage.
 ifdef ALT_GCC29_COMPILER_PATH
diff --git a/make/common/shared/Defs-macosx.gmk b/make/common/shared/Defs-macosx.gmk
--- jdk8/jdk/make/common/shared/Defs-macosx.gmk
+++ jdk8/jdk/make/common/shared/Defs-macosx.gmk
@@ -143,6 +143,11 @@
   _CUPS_HEADERS_PATH=$(PACKAGE_PATH)/include
 endif
 
+# Set ZLIB_LIBS if not already set
+ifeq ("$(ZLIB_LIBS)", "")
+  ZLIB_LIBS=-lz
+endif
+
 # Import JDK images allow for partial builds, components not built are
 #    imported (or copied from) these import areas when needed.
 
diff --git a/make/common/shared/Defs-solaris.gmk b/make/common/shared/Defs-solaris.gmk
--- jdk8/jdk/make/common/shared/Defs-solaris.gmk
+++ jdk8/jdk/make/common/shared/Defs-solaris.gmk
@@ -140,6 +140,11 @@
 
 _CUPS_HEADERS_PATH=/opt/sfw/cups/include
 
+# Set ZLIB_LIBS if not already set
+ifeq ("$(ZLIB_LIBS)", "")
+  ZLIB_LIBS=-lz
+endif
+
 # Import JDK images allow for partial builds, components not built are
 #    imported (or copied from) these import areas when needed.
 
diff --git a/make/java/jli/Makefile b/make/java/jli/Makefile
--- jdk8/jdk/make/java/jli/Makefile
+++ jdk8/jdk/make/java/jli/Makefile
@@ -46,6 +46,8 @@
 
 ifneq ($(SYSTEM_ZLIB),true)
   ZIP_SRC = $(SHARE_SRC)/native/java/util/zip/zlib-$(ZLIB_VERSION)
+else # SYSTEM_ZLIB
+  OTHER_CFLAGS += $(ZLIB_CFLAGS)
 endif #SYSTEM_ZLIB
 LAUNCHER_SHARE_SRC = $(SHARE_SRC)/bin
 
@@ -162,7 +164,7 @@
 ifneq ($(SYSTEM_ZLIB),true)
   OTHER_INCLUDES += -I$(ZIP_SRC)
 else # !SYSTEM_ZLIB
-  LDLIBS += -lz
+  LDLIBS += $(ZLIB_LIBS)
 endif # SYSTEM_ZLIB
 
 #
diff --git a/make/java/zip/Makefile b/make/java/zip/Makefile
--- jdk8/jdk/make/java/zip/Makefile
+++ jdk8/jdk/make/java/zip/Makefile
@@ -56,6 +56,10 @@
   endif
 endif
 
+ifeq ($(SYSTEM_ZLIB),true)
+  OTHER_CFLAGS += $(ZLIB_CFLAGS)
+endif
+
 #
 # Library to compile.
 #
@@ -90,7 +94,7 @@
 # Link to JVM library for JVM_Zip* functions
 #
 ifeq ($(SYSTEM_ZLIB),true)
-OTHER_LDLIBS = -lz
+OTHER_LDLIBS = $(ZLIB_LIBS)
 else
 OTHER_LDLIBS = $(JVMLIB)
 endif
diff --git a/make/jdk_generic_profile.sh b/make/jdk_generic_profile.sh
--- jdk8/jdk/make/jdk_generic_profile.sh
+++ jdk8/jdk/make/jdk_generic_profile.sh
@@ -378,3 +378,22 @@
     export LLVM_LIBS
   fi
 fi
+
+# Export variables for system zlib
+# ZLIB_CFLAGS and ZLIB_LIBS tell the compiler how to compile and
+# link against zlib
+pkgconfig=$(which pkg-config 2>/dev/null)
+if [ -x "${pkgconfig}" ] ; then
+  if [ "${ZLIB_CFLAGS}" = "" ] ; then
+    ZLIB_CFLAGS=$("${pkgconfig}" --cflags zlib)
+  fi
+  if [ "${ZLIB_LIBS}" = "" ] ; then
+    ZLIB_LIBS=$("${pkgconfig}" --libs zlib)
+  fi
+fi
+if [ "${ZLIB_LIBS}" = "" ] ; then
+    ZLIB_LIBS="-lz"
+fi
+export ZLIB_CFLAGS
+export ZLIB_LIBS
+
diff --git a/make/sun/splashscreen/Makefile b/make/sun/splashscreen/Makefile
--- jdk8/jdk/make/sun/splashscreen/Makefile
+++ jdk8/jdk/make/sun/splashscreen/Makefile
@@ -126,7 +126,8 @@
 ifneq ($(SYSTEM_ZLIB),true)
   CPPFLAGS += -I$(SHARE_SRC)/native/java/util/zip/zlib-$(ZLIB_VERSION)
 else
-  OTHER_LDLIBS += -lz
+  OTHER_CFLAGS += $(ZLIB_CFLAGS)
+  OTHER_LDLIBS += $(ZLIB_LIBS)
 endif
 
 # Shun the less than portable MMX assembly code in pnggccrd.c,
diff --git a/src/share/native/com/sun/java/util/jar/pack/defines.h b/src/share/native/com/sun/java/util/jar/pack/defines.h
--- jdk8/jdk/src/share/native/com/sun/java/util/jar/pack/defines.h
+++ jdk8/jdk/src/share/native/com/sun/java/util/jar/pack/defines.h
@@ -93,7 +93,7 @@
 // bytes and byte arrays
 
 typedef unsigned int uint;
-#if !defined(MACOSX) || (defined(MACOSX) && defined(NO_ZLIB))
+#if defined(NO_ZLIB)
 #ifdef _LP64
 typedef unsigned int uLong; // Historical zlib, should be 32-bit.
 #else
diff --git a/src/share/native/java/util/zip/Adler32.c b/src/share/native/java/util/zip/Adler32.c
--- jdk8/jdk/src/share/native/java/util/zip/Adler32.c
+++ jdk8/jdk/src/share/native/java/util/zip/Adler32.c
@@ -29,8 +29,8 @@
 
 #include "jni.h"
 #include "jni_util.h"
-#include "zlib.h"
 #include "jlong.h"
+#include <zlib.h>
 
 #include "java_util_zip_Adler32.h"
 
diff --git a/src/share/native/java/util/zip/CRC32.c b/src/share/native/java/util/zip/CRC32.c
--- jdk8/jdk/src/share/native/java/util/zip/CRC32.c
+++ jdk8/jdk/src/share/native/java/util/zip/CRC32.c
@@ -29,7 +29,7 @@
 
 #include "jni.h"
 #include "jni_util.h"
-#include "zlib.h"
+#include <zlib.h>
 
 #include "java_util_zip_CRC32.h"
 
diff --git a/src/share/native/java/util/zip/Deflater.c b/src/share/native/java/util/zip/Deflater.c
--- jdk8/jdk/src/share/native/java/util/zip/Deflater.c
+++ jdk8/jdk/src/share/native/java/util/zip/Deflater.c
@@ -32,7 +32,7 @@
 #include "jlong.h"
 #include "jni.h"
 #include "jni_util.h"
-#include "zlib.h"
+#include <zlib.h>
 
 #include "java_util_zip_Deflater.h"
 
diff --git a/src/share/native/java/util/zip/Inflater.c b/src/share/native/java/util/zip/Inflater.c
--- jdk8/jdk/src/share/native/java/util/zip/Inflater.c
+++ jdk8/jdk/src/share/native/java/util/zip/Inflater.c
@@ -35,7 +35,7 @@
 #include "jni.h"
 #include "jvm.h"
 #include "jni_util.h"
-#include "zlib.h"
+#include <zlib.h>
 #include "java_util_zip_Inflater.h"
 
 #define ThrowDataFormatException(env, msg) \
diff --git a/src/share/native/java/util/zip/zip_util.c b/src/share/native/java/util/zip/zip_util.c
--- jdk8/jdk/src/share/native/java/util/zip/zip_util.c
+++ jdk8/jdk/src/share/native/java/util/zip/zip_util.c
@@ -44,7 +44,7 @@
 #include "io_util.h"
 #include "io_util_md.h"
 #include "zip_util.h"
-#include "zlib.h"
+#include <zlib.h>
 
 #ifdef _ALLBSD_SOURCE
 #define off64_t off_t
